[FLINK-32560][Scala] Deprecate all Scala API

RyanSkraba opened the pull request:
## What is the purpose of the change

* The entry points for using the Scala API were deprecated in https://github.com/apache/flink/pull/21176.  This PR ensures that all `@Public`, `@PublicEvolving` and `@Experimental` APIs are also marked as `@Deprecated`.
* This PR doesn't have any code changes outside of deprecating these classes. Classes that are *not* scheduled for removal might still need to be migrated.

## Brief change log

* All classes that contain one of those methods are also marked as deprecated with a link to [FLIP-265](https://cwiki.apache.org/confluence/display/FLINK/FLIP-265+Deprecate+and+remove+Scala+API+support).
* The `flink-table`  modules have not been deprecated except for those dealing with a user-facing scala API (`flink-table-api-scala` and `flink-table-api-scala-bridge`)
* The hadoop connector Scala classes were also `@Deprecated`

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): **no**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: **yes**
  - The serializers: **no**
  - The runtime per-record code paths (performance sensitive): **no**
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: **no**
  - The S3 file system connector: **no**

## Documentation

  - Does this pull request introduce a new feature? **no** (but it does prepare for the next major release as described in FLIP-265)

xintongsong commented:
Hi @RyanSkraba, thanks for working on this PR. There are some compiling errors that lead to the CI failure. Could you please take a look?

RyanSkraba commented:
Hello!  This was a mistake on my part, while trying to be consistently put the `@Deprecated` annotation before the `@Public`.  I changed a scala `@deprecated` to a java `@Deprecated`.  As far as I can tell, there's little reason to prefer one over the other in terms of scaladoc or warnings in IntelliJ.  I've pushed the fix!

reswqa commented on flink-connectors/flink-hadoop-compatibility/src/main/scala/org/apache/flink/api/scala/hadoop/mapred/HadoopInputFormat.scala:
```diff
@@ -22,6 +22,17 @@ import org.apache.flink.api.java.hadoop.mapred.HadoopInputFormatBase
 
+/**
+ * @deprecated
+ *   All Flink Scala APIs are deprecated and will be removed in a future Flink version version. You
```
```suggestion
 *   All Flink Scala APIs are deprecated and will be removed in a future Flink major version. You
```

RyanSkraba replied:
Oh my -- I'll fix all of these, of course.  Good catch on this!

reswqa reviewed (commented):
Thanks @RyanSkraba for your efforts. I just took a quick glance and would find some free time to carefully review all the parts.

RyanSkraba commented:
Some quick questions to consider here! There are some classes without a `@Public` or another visibility annotation (like the entire `flink-cep-scala` module).  Do you think it's worthwhile giving them the same `@Deprecated` treatment?

How about the .java classes such as [ScalaAggregateOperator.java](https://github.com/apache/flink/blob/c6997c97c575d334679915c328792b8a3067cfb5/flink-scala/src/main/java/org/apache/flink/api/scala/operators/ScalaAggregateOperator.java#L55C1-L55C1) in the `flink-scala` and `flink-streaming-scala` modules? 

I'm looking at migrating away from some of the newly deprecated methods as a next step (but this can be done in a later logical step after the API is correctly annotated).

xintongsong commented:
@flinkbot run azure

reswqa commented on flink-scala/src/main/java/org/apache/flink/api/scala/typeutils/ScalaOptionSerializerSnapshot.java:
```diff
@@ -24,7 +24,17 @@
+ *     FLIP-265 Deprecate and remove Scala API support</a>
+ */
+@Deprecated
 public final class ScalaOptionSerializerSnapshot<E>
```
I'm not sure if this and other `SerializerSnapshot` should be treat as public API.

RyanSkraba replied:
Yeah, I think they all should have been `@Internal` -- my reasoning was that removing a `@deprecated` annotation should probably always safe, even between minor releases.  Would you prefer these to be removed?

reswqa replied:
IMO, `@Deprecated` is only marked for public api. So I prefer these to be removed.

reswqa reviewed (commented):
Thanks for the update, this change overall looks good to me. I have left some comments.

RyanSkraba commented on flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/function/RichProcessWindowFunction.scala:
```diff
@@ -32,8 +32,16 @@ import org.apache.flink.streaming.api.windowing.windows.Window
+ *   FLIP-265 Deprecate and remove Scala API support</a>
  */
-@Public
 @deprecated("use [[ProcessWindowFunction]] instead")
```
Thanks for the review!  It looks like we currently mix these two in our scala code.

I took a quick look and it appears that:

1. IntelliJ works well for marking deprecated methods in the code regardless of which annotation is used, but
2. the generated scaladoc requires the `@deprecated` annotation on methods.

I propose fixing it throughout all scala API code to use `@deprecated`.

RyanSkraba replied:
OK, this is a bit irritating:

1. Using the Java `@Deprecated` on a class takes the comment from the javadoc and applies it to the class, while scala's `@deprecated` requires the message to be in the annotation.
2. Using the scala `@deprecated("All Flink Scala APIs are deprecated...")` deprecates the class, but not the constructor.  It's correctly barred out when using the Java annotation.

These apply to the scaladoc only -- I suspect that we can live with either version, in principle.

On the other hand, the spotless formatting of the FLIP-165 link breaks in the scaladoc and should be fixed.  You can see this [here](https://nightlies.apache.org/flink/flink-docs-release-1.17/api/scala/org/apache/flink/api/scala/index.html).  I'll replace it with an ASF shortlink.

reswqa replied:
> 1.Using the Java @Deprecated on a class takes the comment from the javadoc and applies it to the class, while scala's @deprecated requires the message to be in the annotation.
2.Using the scala @deprecated("All Flink Scala APIs are deprecated...") deprecates the class, but not the constructor. It's correctly barred out when using the Java annotation.

Does this mean that at the class level, using Java annotation is better? What problems will it bring if we all use Java's `@Deprecated` for `.scala` file? Is it the generated scala doc cannot recognize the `@Deprecated` annotation on method?

twalthr reviewed (commented):
TBH I'm questioning this PR. We only deprecated the entry points to the API because we wanted to keep the API usable while it still exists. Nobody wants to develop in an API where every single method is deprecated. I'm fine maybe with class-level deprecations if we need more of those. But spamming the codebase with deprecations while everyones knows that the entire API will be removed, is not very helpful. It will also spam CI/CD logs with hundreds of warnings.

RyanSkraba commented:
Just to summarize (the main discussion is on the JIRA):

* Rebased and squashed all commits
* Only top level classes are `@deprecated` to ensure that the dev has the info about removing these classes and modules in 2.0 (but the entrypoints to the Scala API have been deprecated since 1.17.0)
* The scala `@deprecation` is preferred, I didn't search and replace existing `@Deprecation` annotations.  These are largely equivalent.

The original (overly annotated) PR is temporarily available at https://github.com/apache/flink/compare/master...RyanSkraba:flink:rskraba/FLINK-32560-deprecate-scala-original

xintongsong reviewed (commented):
Thanks @RyanSkraba. LGTM.

Now my only concern is regarding the CEP classes. I'll try to reach out to @dianfu offline.

dianfu commented:
Per my understanding, all the classes under flink-cep-scala module should be API classes. Currently most API classes under Flink CEP are not annotated as Public / PublicEvolving, have created https://issues.apache.org/jira/browse/FLINK-32724 to track this~

xintongsong commented:
Thanks, @dianfu.

Then I'd suggest to also annotate classes in `flink-cep-scala` as deprecated in this PR, despite they are not annotated as `Public / PublicEvloving`. Because deciding which classes in the module should be `Public` and which should be `PublicEvolving` requires further discussion, which is out of the scope of this PR.

WDYT?

reswqa commented:
Thanks @RyanSkraba, this changes overall looks good to me except for `flink-cep-scala` part. According to @dianfu and @xintongsong 's comment, would you mind also annotate classes in this module as deprecated at the same time?

xintongsong reviewed (approved):
@RyanSkraba, thanks for you contribution. LGTM. Merging.
